docs/proposals/08-shoot-apiserver-via-sni.md

```diff
@@ -0,0 +1,396 @@
+                                                        |
+                                                        |
+                                                        |
+   +---------------------+                   +---------------------+
+   |                     |                   |                     |
+   |  Pod talking to     |                   |                     |
+   |  the kubernetes     |                   |       Proxy         |
+   |  service            +------------------>+  No TLS termination |
+   |                     |                   |                     |
+   |                     |                   |                     |
+   +---------------------+                   +---------------------+
+```
+
+Multiple OSS solutions can be used:
+
+- https://github.com/ginuerzh/gost (various transport options)
+- https://github.com/jpillora/chisel
+- https://github.com/xtaci/kcptun (transport is UDP-based)
+- OpenVPN is discarded as the performance is abysmal
+
+`chisel` is the fastest and offers the least latency when testing under different conditions. The architecture is simple:
+
+![overview](https://docs.google.com/drawings/d/1p53VWxzGNfy8rjr-mW8pvisJmhkoLl82vAgctO_6f1w/pub?w=960&h=720)
+
+For the Gardener implementation, the `clisel-client` will run as a `DaemonSet` on the `Shoot` cluster and a dedicated `chisel-server` is provisioned for each `Shoot` control plane. To reduce the amount of records we create, it listens on a dedicated port `9443`.
+
+An updated version of the previous SNI configuration:
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: kube-apiserver-gateway
+  namespace: shoot-namespace
+spec:
+  selector:
+    istio: ingressgateway
+  servers:
+  - port:
+      number: 443
+      name: tls
+      protocol: TLS
+    tls:
+      mode: PASSTHROUGH
+    hosts:
+    - api.<shoot>.<seed>.<domain>
+    - api.internal.<shoot>.<seed>.<domain>
+  - port:
+      number: 9443
+      name: tls
+      protocol: TLS
+    tls:
+      mode: PASSTHROUGH
+    hosts:
+    - api.internal.<shoot>.<seed>.<domain>
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: kube-apiserver
+  namespace: shoot-namespace
+spec:
+  hosts:
+  - api.<shoot>.<seed>.<domain>
+  - api.internal.<shoot>.<seed>.<domain>
+  gateways:
+  - kube-apiserver-gateway
+  tls:
+  - match:
+    - port: 443
+      sniHosts:
+      - api.<shoot>.<seed>.<domain>
+      - api.internal.<shoot>.<seed>.<domain>
+    route:
+    - destination:
+        host: kube-apiserver.shoot-namespace.svc.cluster.local
+        port:
+          number: 443
+  - match:
+    - port: 9443
+      sniHosts:
+      - api.internal.<shoot>.<seed>.<domain>
+    route:
+    - destination:
+        host: proxy.shoot-namespace.svc.cluster.local
+        port:
+          number: 443
+```
+
+Chisel works by creating logical TCP connections via SSH, but it uses only one TCP connection per client/Shoot Node.
```

Comment:
Can you elaborate why? is this for load-balancing across L4? in that case, is the timeout for each connection configurable (how do we maintain timeouts across different connections)?

Reply:
> Can you elaborate why? is this for load-balancing across L4?

The main idea is to improve performance.

> in that case, is the timeout for each connection configurable 

The ssh client has a 30sec hardcoded timeout https://github.com/jpillora/chisel/blob/fb7357509aa28a62c6e87d4e4cba958ca66a08ab/client/client.go#L107 and on the server side we are looking at 10 seconds for establishing a connection: https://github.com/jpillora/chisel/blob/4a85f97336424cc8b734fb5e98afbc318fcf1a50/server/handler.go#L75